Add vector Dirichlet BCs and coupled vector gradient methods

lindsayad opened the pull request:
- Implement Dirichlet boundary conditions for `LAGRANGE_VEC` variables
- Add `coupledVectorGradient` and corresponding `old` and `older` methods for accessing gradients of vector variables (rank two tensors)
- Adds test which tests both vector `Dirichlet` and `coupledVectorGradient` capabilities
- Adds new `dofValues` variable methods that replace the functionality of the old `nodalValue` methods
- The `nodalValue` methods stick around, but they now just return a single `Real` or a `RealVectorValue` for vector variables; e.g. we try to be more faithful to the method name

Closes #11150

lindsayad committed:
Add coupledVectorGradient method. Closes #11150

lindsayad commented:
@dschwen If you have a quick test in your head, feel free to push to this branch. Otherwise I can create one

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/171996/) on 171766e wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11151/site)

This comment will be updated on new commits.

dschwen reviewed (dismissed):
LGTM. Are you also doing the _old_ and _older_ methods?

permcody commented:
Let's not merge this in until we get a simple test add. Don't worry about actual physics. Just set up a few fields that have a some sort of shape and couple to them to verify this works.

lindsayad committed:
Add VectorNodalBC

lindsayad committed:
Implement NodalBCBase class

lindsayad committed:
Add old gradient methods and corresponding test.

lindsayad committed:
Add docs

dschwen reviewed (dismissed):
A smorgasbord of changes! Tests are there now and additional features are as well...

permcody commented on framework/include/bcs/LagrangeVecFunctionDirichletBC.h:
```diff
@@ -0,0 +1,48 @@
+  LagrangeVecFunctionDirichletBC(const InputParameters & parameters);
+
+protected:
+  virtual Real computeQpResidual(size_t comp) override;
```
I'd prefer to use the C++ version for our new APIs. Let's not mix C and C++:
https://stackoverflow.com/questions/5813700/difference-between-size-t-and-stdsize-t

This is going to apply all over the place in this PR... Take a look over everything and and update appropriately.

permcody replied:
I know this seems like a big pain but note that according to the standard. You are only guaranteed to get the non-namespaced version if you include `stddef.h`. In C++, you would include `cstddef` and get only the namespaced version. Now in practice, we are going to always get both because real headers are almost never 100% clean. I think it's worth doing right.

lindsayad committed:
Make nodalValue mean what it says and add dofValues methods for moose variables

- nodalValue now only returns a single value (or a 3-length vector for LAGRANGE_VEC
  variables)
- dofValues methods replace nodalValue calls in places where the user expects to receive
  an array of values corresponding to the degree of freedom values
- `dof_u` members in MooseVariable* are renamed to `dof_values` to hopefully be a bit
  less confusing
- restore VectorNodalBC computeQpResidual to return a RealVectorValue

permcody commented:
Removing the "Ready to Merge" tag. I think we decided we don't want to expose any kind of size_t to the user.

lindsayad commented:
@friedmud In nodal computing objects (with the exception of the new `VectorNodalBC` which uses the "new" `nodalValue` method), I have switched the construction of the `_u` member variable to use `dofValues` instead of `nodalValue` since the latter now supplies only a single value and subsequently indexing with `_qp` would cause errors. 

Eventually per our discussion, I think we want to switch nodal compute objects to use the new `nodalValue` and remove the `_qp` indexing. That will require breaking/fixing a lot of apps. I would prefer not to do that in this PR (which was meant to originally be a small change) but can if you think it should happen here.

friedmud commented on framework/include/variables/MooseVariableFE.h:
```diff
@@ -334,17 +371,17 @@ class MooseVariableFE : public MooseVariableBase
-  bool _need_dof_u_older;
-  bool _need_dof_u_previous_nl;
-  bool _need_dof_u_dot;
+  bool _need_dof_values;
```
Thanks changing to `dof_values`

friedmud commented on framework/src/auxkernels/AuxKernel.C:
```diff
@@ -98,9 +98,9 @@ AuxKernel::AuxKernel(const InputParameters & parameters)
-    _u(_nodal ? _var.nodalValue() : _var.sln()),
-    _u_old(_nodal ? _var.nodalValueOld() : _var.slnOld()),
-    _u_older(_nodal ? _var.nodalValueOlder() : _var.slnOlder()),
+    _u(_nodal ? _var.dofValues() : _var.sln()),
```
👍 This looks better now - at least it reflects what's actually happening

friedmud commented on framework/src/bcs/LagrangeVecDirichletBC.C:
```diff
@@ -30,8 +30,8 @@ LagrangeVecDirichletBC::LagrangeVecDirichletBC(const InputParameters & parameter
+LagrangeVecDirichletBC::computeQpResidual()
 {
-  return _u[comp] - _values(comp);
+  return _u - _values;
```
👍

friedmud commented on framework/src/kernels/MassLumpedTimeDerivative.C:
```diff
@@ -30,7 +30,7 @@ validParams<MassLumpedTimeDerivative>()
 
 MassLumpedTimeDerivative::MassLumpedTimeDerivative(const InputParameters & parameters)
-  : TimeKernel(parameters), _u_dot_nodal(_var.nodalValueDot())
+  : TimeKernel(parameters), _u_dot_nodal(_var.dofValuesDot())
```
Again - this looks much better.  In the short term `_u_dot_nodal` is fine - but hopefully we can change it in the future.

friedmud reviewed (approved):
This looks really good - I think the changes are definitely positive.  Thanks for working on this!
